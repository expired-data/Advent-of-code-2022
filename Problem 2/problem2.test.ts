import {
  isMyWin,
  isDraw,
  scoreRound,
  parseRounds,
  solvePart1,
  Choice,
  Round,
} from "./problem2";
import fs from "fs";
import path from "path";

const file = path.join(__dirname, "./", "example-input.txt");
const exampleData: string = fs.readFileSync(file, "utf8");

const allChoices = [Choice.Rock, Choice.Paper, Choice.Scissors];
const allPairs: Round[] = allChoices.flatMap((choice: Choice): Round[] =>
  allChoices.map((choice2: Choice): Round => [choice, choice2])
);

describe("isMyWin", () => {
  test("should win 3 times across allPairs", () => {
    const rounds = allPairs.filter((round: Round): boolean => isMyWin(round));

    expect(rounds.length).toBe(3);
  });
});

describe("isDraw", () => {
  test("should draw only when left and right are equal", () => {
    allPairs.forEach(([elfChoice, myChoice]: Round): void => {
      if (elfChoice == myChoice) {
        expect(isDraw([elfChoice, myChoice])).toBe(true);
      } else {
        expect(isDraw([elfChoice, myChoice])).toBe(false);
      }
    });
  });
});

describe("scoreRound", () => {
  test("round scores are valid for all values", () => {
    const roundScores = allPairs.map((round: Round): number =>
      scoreRound(round)
    );

    expect(roundScores).toMatchSnapshot();
  });
});

describe("parseRounds", () => {
  test("can parse example rounds", () => {
    expect(parseRounds(exampleData)).toEqual([
      [Choice.Rock, Choice.Paper],
      [Choice.Paper, Choice.Rock],
      [Choice.Scissors, Choice.Scissors],
    ]);
  });
});

describe("solvePart1", () => {
  test("can solve example rounds", () => {
    expect(solvePart1(exampleData)).toBe(15);
  });
});
