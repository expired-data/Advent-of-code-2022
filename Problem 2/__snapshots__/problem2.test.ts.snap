// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scoreRound round scores are valid for all values 1`] = `
[
  4,
  8,
  3,
  1,
  5,
  9,
  7,
  2,
  6,
]
`;
